src/annotator/guest.js

```diff
@@ -465,21 +469,35 @@ export default class Guest {
     removeAllHighlights(this.element);
 
     this._integration.destroy();
-    this._sidebarRPC.destroy();
   }
 
   /**
-   * Notify the host frame that the guest is being unloaded.
+   * Notify other frames that the guest is being unloaded.
+   *
+   * In order to work around a bug in Safari 15 and below [1], this is done by
+   * first sending ports to the host frame and then, from the host frame,
+   * sending messages on the ports.
    *
-   * The host frame in turn notifies the sidebar app that the guest has gone away.
+   * [1] https://bugs.webkit.org/show_bug.cgi?id=231167.
    */
-  _notifyGuestUnload() {
-    const message = {
-      type: 'hypothesisGuestUnloaded',
-      frameIdentifier: this._frameIdentifier,
-    };
+  _sendUnloadNotification() {
+    const ports = [];
+
+    const sidebarPort = this._sidebarRPC.disconnect();
+    if (sidebarPort) {
+      ports.push(sidebarPort);
+    }
 
-    this._hostFrame.postMessage(message, '*');
+    const hostPort = this._hostRPC.disconnect();
```

Comment:
I have been thinking about what would happen if the guest frame is unloaded while there is an in-flight host discovery process. In other words, what would happened if  `this._hostRPC.disconnect()` is called before `this._hostRPC.connect(port)`. 

In that scenario, `this._hostRPC.disconnect()` will return `null` and it will fail to notify the host or sidebar frames of the unloading.

Reply:
> In that scenario, this._hostRPC.disconnect() will return null and it will fail to notify the host or sidebar frames of the unloading.

This is true. Sadly the Channel Messaging API lacks a built-in solution for this. In the case of the guest-sidebar connection, we can use the lack of `documentInfoChanged` message as an indicator. I think it would make sense to have a generic solution in the `Port*` classes.

There might also be scenarios where the frame is unloaded without having the ability to send a notification on the port. I'm not sure yet if that can happen in any browsers or whether browsers will guarantee to unload entire tabs (eg. to save memory) as an atomic unit.